src/main/java/com/diffmin/patch/Generation.java

```diff
@@ -0,0 +1,102 @@
+package com.diffmin.patch;
+
+import com.diffmin.SpoonMapping;
+import com.diffmin.util.Pair;
+import com.github.gumtreediff.actions.model.Delete;
+import com.github.gumtreediff.actions.model.Insert;
+import com.github.gumtreediff.actions.model.Update;
+import gumtree.spoon.diff.Diff;
+import gumtree.spoon.diff.operations.Operation;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.IntStream;
+import org.apache.commons.lang3.tuple.ImmutableTriple;
+import spoon.reflect.code.CtInvocation;
+import spoon.reflect.code.CtStatementList;
+import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtExecutable;
+
+/** Class for generating patches. */
+public class Generation {
+    public final Set<CtElement> deletePatches = new HashSet<>();
+    public final Set<Pair<CtElement, CtElement>> updatePatches = new HashSet<>();
+    public final Set<ImmutableTriple<Integer, CtElement, CtElement>> insertPatches =
+            new HashSet<>();
```

Comment:
Two problems here:

### `HashSet` for order-dependent collections

Using a `HashSet` for insert patches is probably not a good idea. If I'm not mistaken, insert patches are dependent on the order of operations, as the computed index in a given list is only correct if all preceding insertions have already been performed. Thus, if you really want a set here, you should use one that preserves insertion order (a property that `HashSet` does not have).

However, you shouldn't use sets here. For `deletePatches`, you can have incorrect behavior as two equal nodes (say `int a = 2` appearing in two different places) would interfere with each other (only one can be added to the set). Thus, you might lose deletions.

For `updatePatches`, your `Pair` class doesn't implement `equals` and `hashCode`, and so its set properties in a `HashSet` will simply be those of `Object` (compare by reference). Ironically, this is the only one that will actually work as intended!

Finally, for `insertPatches`, you have the same problem as for `deletePatches`, assuming that `ImmutableTriple` has a reasonable `equals` and `hashCode` implementation.

To summarize: I would just use lists here.

### Public mutable fields

Public, mutable fields is a big no-no. When you mark a collection as final, that only applies to the reference in the field, and not what it points to. For example, I can't do this:

```java
Generation generation = new Generation();
generation.deletePatches = new HashSet<>(); // error, deletePatches is final
```

but I can do this:

```java
Generation generation = new Generation();
generation.deletePatches.add(someCtElement); // this is fine
```

Not only does it allow me to see the internals of the class from the outside, it allows me to modify them!

Reply:
> a property that HashSet does not have

Oh, I thought `HashSet` preserves order while maintaing the uniqueness.

> For deletePatches, you can have incorrect behavior as two equal nodes (say int a = 2 appearing in two different places) would interfere with each other (only one can be added to the set). Thus, you might lose deletions.

I put some thought here and then I realised that the `CtElement` in the delete patch will have an attribute `parent` which will definitely differ so same statements can co-exist if the previously stated condition is met.

Reply:
> I put some thought here and then I realised that the CtElement in the delete patch will have an attribute parent which will definitely differ so same statements can co-exist if the previously stated condition is met.

This doesn't work for two reasons.

1. `parent` is not part of the equality check of a `CtElement`. The semantics of `CtElement.equals(CtElement)` is to check if they two _subtrees rooted in those elements_ are equal.
2. Even if `parent` was part of the equality check, what about this (`a++` nodes are identical):

```java
int a = 2;
a++;
a++;
```

Reply:
Oh, okay. I didn't think through. I did not know if `equals` was overridden. I assumed it would have a default behaviour. 

> Even if parent was part of the equality check, what about this (a++ nodes are identical):

Great example! Thanks :)

Reply:
> Oh, okay. I didn't think through. I did not know if equals was overridden. I assumed it would have a default behaviour.

Default behavior for equals is to compare by reference (see [Object.equals()](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#equals-java.lang.Object-). So, for _any_ class that requires true equality comparison, one must override equals (which also entails overriding hashCode). Perhaps we should discuss this during the meeting, it's one of those fundamental Java things that one must be familiar with.